client/app/bundles/course/assessment/submission/containers/VoiceResponseAnswer.jsx

```diff
@@ -0,0 +1,209 @@
+import React, { Component } from 'react';
+import { connect } from 'react-redux';
+import PropTypes from 'prop-types';
+import { Field } from 'redux-form';
+import { injectIntl, intlShape, defineMessages, FormattedMessage } from 'react-intl';
+import FlatButton from 'material-ui/FlatButton';
+import MicIcon from 'material-ui/svg-icons/av/mic';
+import StopIcon from 'material-ui/svg-icons/av/stop';
+import SingleFileInput from 'lib/components/redux-form/SingleFileInput';
+import recorderHelper from '../../utils/recorderHelper';
+import sharedConstants from '../../../../../lib/constants/sharedConstants';
+import {
+  setRecording,
+  setNotRecording,
+  recorderComponentMount,
+  recorderComponentUnmount,
+} from '../actions/index';
+
+const translations = defineMessages({
+  startRecording: {
+    id: 'course.assessment.submission.answer.startRecording',
+    defaultMessage: 'Start Recording',
+  },
+  stopRecording: {
+    id: 'course.assessment.submission.answer.stopRecording',
+    defaultMessage: 'Stop Recording',
+  },
+  chooseVoiceFileExplain: {
+    id: 'course.assessment.submission.answer.chooseVoiceFileExplain',
+    defaultMessage: 'Drag your audio file here, or click to select an audio file. \
+                     Only wav and mp3 formats are supported. Alternatively, you may use the \
+                     recorder below to record your response',
+  },
+});
+
+const styles = {
+  fileInputWrapper: {
+    width: '60%',
+  },
+  singleFileInputChildrenWrapper: {
+    width: '100%',
+    height: '100%',
+    display: 'table',
+  },
+  singleFileInputChildren: {
+    display: 'table-cell',
+    verticalAlign: 'middle',
+  },
+};
+
+class VoiceResponseAnswer extends Component {
```

Comment:
This component seems like it is trying to handle quite a few things - would be it possible to break it up into 2, similar to `SingleAudioFileInput` (which includes the file input and recorder), and `VoiceResponseAnswer` which deals more with the answer and forms?

Reply:
I am ok if you feel that it is okay to leave it here, but I'll just state why I suggested it. 

My concern is not really about the length of the component, but rather the functions the component has to do. As of now, because there are no tests, this is easy to construct (to work). Once we have tests, this component has many things that we have to test. Future refactorings (eg. replacing of recorder js with another library) would be harder as one would have to understand this entire component before making and changes too, and same for any refactoring to SingleFileInput (eg. if we want to unify the FileInput components across the codebase).